README.md
# Workbench Images

*[Changelog](CHANGELOG.md)*

Various **Workbench and Runtime images** to use with [Open Data Hub](http://opendatahub.io/) (ODH) or [Red Hat OpenShift Data Science](https://www.redhat.com/en/technologies/cloud-computing/openshift/openshift-data-science) (RHODS).

The Workbench and Runtime images are built in a modular way, allowing for full flexibility in the packages and applications (**Datascience**, **PyTorch**, **Tensorflow**, **Langchain**,...) or IDEs (**Jupyter**, **VSCode**, **RStudio**, **Langflow**,...) you want to use.

With all those possible configurations, only some images are pre-built and published (the most important ones). Those are listed below.

An interactive **Wizard** allows you to also easily create you own image! See [Building an image](#building-an-image).

## Available Images

All images listed below are based on CentOS Stream 9 with Python 3.11 (except the Langflow image based on 3.9 due to compatibility issues).

### Workbench images

Those images can be used directly as Workbenches within ODH or RHODS after importing them (Settings->Notebook images->Import new image).

| Image description | Pull address |
|-------------------|--------------|
|Jupyter Data Science|quay.io/opendatahub-contrib/workbench-images:jupyter-datascience-c9s-py311_2023c_latest|
|VSCode Data Science|quay.io/opendatahub-contrib/workbench-images:vscode-datascience-c9s-py311_2023c_latest|
|RStudio|quay.io/opendatahub-contrib/workbench-images:rstudio-r-c9s-py311_2023c_latest|
|Jupyter PyTorch (CUDA)|quay.io/opendatahub-contrib/workbench-images:cuda-jupyter-pytorch-c9s-py311_2023c_latest|
|VSCode PyTorch (CUDA)|quay.io/opendatahub-contrib/workbench-images:cuda-vscode-pytorch-c9s-py311_2023c_latest|
|Jupyter Tensorflow (CUDA)|quay.io/opendatahub-contrib/workbench-images:cuda-jupyter-tensorflow-c9s-py311_2023c_latest|
|Jupyter Langchain (CUDA)|quay.io/opendatahub-contrib/workbench-images:cuda-jupyter-langchain-c9s-py311_2023c_latest|
|Langflow (CUDA)|quay.io/opendatahub-contrib/workbench-images:cuda-langflow-langflow-c9s-py39_2023c_latest|

### Runtime images

Those images can be used in a Data Science Pipeline. They contain the exact same set of libraries as workbench images, except the IDE part.

| Image description | Pull address |
|-------------------|--------------|
|Runtime Data Science|quay.io/opendatahub-contrib/workbench-images:runtime-datascience-c9s-py311_2023c_latest|
|Runtime Data Science (CUDA)|quay.io/opendatahub-contrib/workbench-images:cuda-runtime-datascience-c9s-py311_2023c_latest|
|Runtime PyTorch (CUDA)|quay.io/opendatahub-contrib/workbench-images:cuda-runtime-pytorch-c9s-py311_2023c_latest|
|Runtime Tensorflow (CUDA)|quay.io/opendatahub-contrib/workbench-images:cuda-runtime-tensorflow-c9s-py39_2023c_latest|
|Runtime R|quay.io/opendatahub-contrib/workbench-images:runtime-r-c9s-py311_2023c_latest|
|Runtime Spark|quay.io/opendatahub-contrib/workbench-images:runtime-spark-c9s-py311_2023c_latest|

## Detailed Images Content

Images consist of:

- A base OS image
- Some base OS packages
- A bundle of applications and libraries
- An IDE (or not in the case of Runtime images)

So with different sets of OS packages, bundles and IDEs to choose from, you can create the image most suited to your need!

### Base (OS)

All images are based on CentOS Stream 9. This allows to install applications and libraries that would not be available on UBI-based images. The images could also be easily ported to RHEL to get full support. Epel and CRB repos are enabled, as well as rpmfusion and other proprietary repos from Microsoft, Cisco and MongoDB.

### OS Packages

The base images include the following main libraries out of the box:

- FFmpeg
- GStreamer + various codecs
- OpenH264
- MongoDB Client
- MSSQL Client
- MySQL Client
- PostgreSQL Client
- UnixODBC
- Git LFS

**Full list**: diffutils, ffmpeg, git-lfs, gstreamer1, gstreamer1-libav, gstreamer1-plugin-openh264, gstreamer1-plugins-base, gstreamer1-plugins-base-tools, hdf5, jq, lame, lcms2, libsndfile, llvm, llvm-devel, lz4, mongocli, mssql-tools18, mysql, nano, openh264, postgresql, tk, unixODBC, unixODBC-devel, zstd.

### Python Version

Whenever possible, applications/libraries bundles are available for both Python 3.9 and Python 3.11.

### Bundles

Bundles are sets of applications and libraries available to build the different images, depending of the functionalities you want. The following bundles are available in the latest release (2023c).

| Bundle name | Description | Included Libraries/Application |
|-------------------|--------------|--------------|
|Minimal|Minimal set of Python packages|pip~=23.2.1, setuptools~=68.1.2, wheel~=0.41.2|
|Data Science|Standard set of Data Science tools|Minimal + boto3~=1.28.40, codeflare-sdk~=0.7.1, kafka-python~=2.0.2, kubernetes~=25.3.0, matplotlib~=3.7.2, numpy~=1.24.4, pandas~=2.1.0, plotly~=5.16.1, scikit-learn~=1.3.0, scipy~=1.11.2, skl2onnx~=1.15.0, pyodbc~=4.0.39, pymongo~=4.5.0, psycopg~=3.1.10, mysql-connector-python~=8.0.33|
|PyTorch|PyTorch 2.0.1 + TorchVision|Data Science + torch==2.0.1+cu118, torchvision==0.15.2+cu118, tensorboard~=2.14.0|
|Tensorflow|Tensorflow 2.13.0 + TF2Onnx|Data Science + tensorflow~=2.13.0, tensorboard~=2.13.0, tf2onnx~=1.15.1|
|LangChain|PyTorch 2.0.1 + LangChain 0.0.285|PyTorch + langchain==0.0.285 (extras=["llms"]), text-generation = "~=0.6.0"|
|Langflow (Python 3.9 only)|PyTorch 2.0.1 + LangChain 0.0.256 + LangFlow 0.4.18|PyTorch + langchain==0.0.256 (extras=["llms"]), langflow==0.4.18, text-generation = "~=0.6.0"|
|OptaPy|OptaPy 9.37.0b0|Data Science + optapy==9.37.0b0 + Java 11|
|R|R 4.3.1|R-core, R-core-devel, R-java, R-Rcpp, R-highlight, R-littler, R-littler-examples|
|Spark|Spark 3.4.1|Data Science + pyspark~=3.4.1 + Spark 3.4.1 + Java 11|

### IDEs

The following IDEs are available:

- Jupyter ~=3.6.5
- VSCode 4.16.1
- RStudio 2023.06.01
- LangFlow 0.4.18 (will only work with the Langflow bundle)

## Building an image

If the pre-built images are not enough, you can easily create your own one with the available bundles and IDEs.

**Requirements**: [Podman](https://podman.io/) and [Skopeo](https://github.com/containers/skopeo/blob/main/install.md) have to be installed on your system to be able to use the wizard.

- Clone this repo.
- Launch the wizard: `./interactive-image-builder.sh`
- Answer a few questions.

Based on your choices, the wizard will create the recipe and give you the instructions on how to build the image.

## Development

For the development of new recipes for Bundles, IDEs,.. please see the [Development](DEVELOPMENT.md) page.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [2023c] - 2023-07-17

- Remove Python 3.8 support
- Added Python 3.11 support
- Switch all images to CentOS Stream 9 (c9s)
- Use new base image [https://quay.io/repository/sclorg/python-39-c9s](https://quay.io/repository/sclorg/python-39-c9s)
- Runtimes set now before workbench images to simplify chaining
- Toolkit installed in CUDA Development images
- Add PyTorch and Tensorflow images because of the new c9s base
- Fix #28
- Updates to apps and packages
  - VSCode: 4.9.1 => 4.16.1
  - R: 4.2.2 => 4.3.1
  - RStudio: 2022.12.0-353 => 2023.06.1-524

## [2023b-patch]

- Fixes on VSCode and RStudio images for bad redirects, integrated in 2023b
- Addition of TrustyAI workbench image, integrated in 2023b
- Addition of runtime images, integrated in 2023b

## [2023b] - 2023-02-01

- Update to the CUDA RStudio and Jupyter + RStudio images to include the CUDA toolkit. This is necessary to install certain packages that are compiled in-place.
- Update to RStudio and VSCode images (standalone versions). They are now fully compatible with the dashboard, so can be used as standard custom images. Idle culling also works with those images.
- Addition of some packages:
  - Minimal:
    - jupyter-server-terminals ~=0.4.4 and jupyter-server ~=2.1.0: this will allow culling to work for terminals once enable on the notebook controller.
- Updates of some packages versions:
  - Minimal:
    - jupyterlab ~=3.5.2 -> ~=3.5.3
    - jupyter-resource-usage: ~=0.6.4 -> ~=0.7.0
  - Datascience:
    - matplotlib ~=3.6.2 -> ~=3.6.3
    - pandas ~=1.5.2 -> ~=1.5.3
    - plotly ~=5.11.0 -> ~=5.13.0
    - scikit-learn ~=1.2.0 -> ~=1.2.1
    - scipy ~=1.9.3 -> ~=1.10.0
    - streamlit ~=1.16.0  -> ~=1.17.0

## [2023a] - 2023-01-02

- Migration of repository to opendatahub-io-contrib organization
- Modification of versioning scheme to allow multiple parallel releases and automated regular updates as described [here](README.md#naming-convention)
- Full refactoring of the repository for clarity
- Adding tests to the Makefiles to validate images
- Release of 2023a, creation of associated support branch

## [0.0.15] - 2022-12-02

- Spark notebook in the `notebook-controller-images` folder. This is a JupyterLab Data Science notebook with Spark 3.3.1 and Hadoop 3.3.4.
- Also includes RBAC and examples for launching Spark workloads directly from the notebook.

## [0.0.14] - 2022-10-23

- VSCode notebook in the `jupyterhub-images` folder. This is a JupyterLab Data Science notebook that also includes VSCode 4.7.1. This image works both with JupyterHub and current ODH/RHODS dashboard version (contrary to the image in `notebook-controller-images` that works only with the Kubeflow Notebook Controller and is not compatible with ODH/RHODS dashboard).

## [0.0.13] - 2022-10-05

- Monai notebook in the `notebook-controller-images` line. As it's not thoroughly tested it's still not officially released
- OptaPy custom image updated with OptaPy 8.23.0a0, minimal-data-science-optapy-notebook:v0.0.2
- All images, including those in the `jupyterhub-images` folder are now compatible with the Kubeflow Notebook Controller

## [0.0.12] - 2022-08-04

### Added

- New organization of images for upcoming Kubeflow Notebook Controller
- Base, Minimal, Data Science, Code-Server and RStudio images available in the new `notebook-controller-images` line

## [0.0.11] - 2022-06-24

### Added

- New notebook image: CUDA 11.6.2 + Monai 0.9.0 + Streamlit 1.10.0 notebook (quay.io/guimou/monai-notebook:0.0.2)

### Modified

- Streamlit image updated to streamlit 1.10.0 - fixes #5
- fix Streamlit launch scripts for directory context - fixes #6
- new streamlit image: quay.io/guimou/generic-data-science-streamlit-notebook:0.0.7

## [0.0.10] - 2022-06-22

### Added

- nbgitpuller added to Streamlit notebook (quay.io/guimou/generic-data-science-streamlit-notebook:0.0.6)

### Modified

- fix yarn cache clean on Streamlit notebook image
- prebuilt-images repos added to CHANGELOG

## [0.0.9] - 2022-06-08

### Added

- Streamlit notebook (quay.io/guimou/generic-data-science-streamlit-notebook:0.0.5)

## [0.0.8] - 2022-03-28

### Added

- Changelog file
- SageMath notebook (quay.io/guimou/minimal-data-science-sagemath-notebook:0.0.1)

## [0.0.7] - 2022-03-25

### Added

- RStudio added to the R notebook (quay.io/guimou/odh-minimal-data-science-r-notebook:0.0.15)

## [0.0.6] - 2022-03-23

### Added

- OptaPy notebook added (quay.io/guimou/minimal-data-science-optapy-notebook:v0.0.1)


interactive-image-builder.sh
#!/bin/bash

set -a

source helper-functions

# Requirements
# Check if skopeo is installed
if ! command -v skopeo &> /dev/null; then
    echo "Error: skopeo is not installed on your system."
    exit 1
fi
# Check if podman is installed
if ! command -v podman &> /dev/null; then
    echo "Error: podman is not installed on your system."
    exit 1
fi

# Intro
clear
echo "###############################"
echo "# The Custom Images Mageforge #"
echo "###############################"
echo
tw_echo "Welcome, brave Container voyager, to the depths of the Custom Images Mageforge, where we conjure containers of power and wisdom! 🔮🐉" 1
tw_echo "You have embarked on a perilous journey through the arcane corridors of containerization." 1
echo
tw_echo "To forge your Container image, you must answer the questions I present to you with wisdom and cunning." 1
tw_echo "Choose your path wisely, and we shall emerge victorious together! 🌀🗝️" 1
echo
tw_echo "Press any key to start this treacherous quest... 🌟" 1
read -n 1 -s

# Promt for release
clear
echo "Release name"
echo "------------"
tw_echo "Which name do you want to give to this release, for example: '2026a' (for default, '2023c', press enter): " 0
read -p "" RELEASE
RELEASE=${RELEASE:-"2023c"}

# Promt for date
clear
echo "Release date"
echo "------------"
tw_echo "Which date do you want to use for this release in YYYYMMDD format (for default '$(date +%Y%m%d)', press enter): " 0
read -p "" DATE
DATE=${DATE:-$(date +%Y%m%d)}

# Promt for base image
clear
echo "Which Base Image do you want to use?"
echo "------------------------------------"
tw_echo "1) Standard image (default, press enter)" 1
tw_echo "2) CUDA image - version 11.8" 1
tw_echo "3) CUDA development image - version 11.8" 1
tw_echo "4) Other" 1

BASE_CHOICE=$(select_value 4)

# Prompt for Python version
clear
echo "Which Python Version do you want to use?"
echo "----------------------------------------"
tw_echo "1) 3.11 (default, press enter)" 1
tw_echo "2) 3.9" 1

PYTHON_CHOICE=$(select_value 2)

case $PYTHON_CHOICE in
  1)
    PYTHON_VERSION="py311"
    ;;
  2)
    PYTHON_VERSION="py39"
    ;;
esac

# Build base image name
case $BASE_CHOICE in
  1)
    BASE_IMAGE="workbench-images:base-c9s-${PYTHON_VERSION}_${RELEASE}_${DATE}"
    ;;
  2)
    BASE_IMAGE="workbench-images:cuda-c9s-${PYTHON_VERSION}-runtime-cudnn_${RELEASE}_${DATE}"
    ;;
  3)
    BASE_IMAGE="workbench-images:cuda-c9s-${PYTHON_VERSION}-devel-cudnn_${RELEASE}_${DATE}"
    ;;
  4)
    tw_echo "Enter the full path/name of your base image (e.g. quay.io/my_repo/my_image:mytag): " 0
    read -p "" BASE_IMAGE
    ;;
esac

# Confirmation of base image
clear
podman images --format "{{.Repository}}:{{.Tag}}" | grep $BASE_IMAGE > /dev/null 2>&1
if [[ $? -eq 0 ]]; then
  tw_echo "The base image that will be used is:" 1
  tw_echo "  $BASE_IMAGE" 1
  tw_echo "It is present on your system, all is good!" 1
else
  tw_echo "The base image you want to use is:" 1
  tw_echo "  $BASE_IMAGE" 1
  echo
  if [[ ! $BASE_CHOICE == 4 ]]; then
    skopeo inspect docker://quay.io/opendatahub-contrib/${BASE_IMAGE} > /dev/null 2>&1
    if [[ $? -eq 0 ]]; then
      tw_echo "I have found this image on quay.io/opendatahub-contrib, all is good!" 1
    else
      skopeo inspect docker://quay.io/opendatahub-contrib/${BASE_IMAGE::-8}latest > /dev/null 2>&1
      if [[ $? -eq 0 ]]; then
        tw_echo "This image could not be found on your system, so I will use the equivalent latest build from quay.io/opendatahub-contrib instead." 1
        echo
        tw_echo "The base image will be:" 1
        tw_echo "  quay.io/opendatahub-contrib/${BASE_IMAGE::-8}latest" 1
        BASE_IMAGE="quay.io/opendatahub-contrib/${BASE_IMAGE::-8}latest"
      else
        tw_echo "Sorry, I could not find this image either locally or its latest version on quay.io/opendatahub-contrib, exiting..." 1
        exit 1
      fi
    fi 
  else
    skopeo inspect docker://$BASE_IMAGE > /dev/null 2>&1
    if [[ $? -eq 0 ]]; then
      tw_echo "This image is not present on your system but I found it in the repo you indicated, so all is good!" 1
    else
      tw_echo "Sorry, I could not find the image you indicated, exiting..." 1
      exit 1
    fi
  fi
fi
tw_echo "Press any key to continue (CTRL+C to end the wizard)..." 0
read -n 1 -s

# Prompt for App Bundle
clear
echo "Which Applications/Libraries bundle do you want to use?"
echo "-------------------------------------------------------"
bundle_folder="./snippets/bundles"
APP_BUNDLE_FOLDER="$(select_subfolder "$bundle_folder" "$PYTHON_VERSION")"
# Remove the leading number and hyphen
APP_BUNDLE_NAME=${APP_BUNDLE_FOLDER#*-}

# Prompt for IDE choice
clear
echo "Which IDE do you want to use?"
echo "------------------------------"
ide_folder="./snippets/ides"
IMAGE_IDE_FOLDER="$(select_subfolder "$ide_folder")"
IMAGE_IDE_NAME=${IMAGE_IDE_FOLDER#*-}

if [[ $IMAGE_IDE_NAME == "runtime" ]]; then
  IMAGE_TYPE="runtime"
else
  IMAGE_TYPE="workbench"
fi

# Build the image
clear
tw_echo "Let's create you image recipe!" 1
echo
tw_echo "In which folder do you want to store this recipe (for default, 'custom-recipes', press enter)?" 1
read -p "" EXPORT_FOLDER
EXPORT_FOLDER=${EXPORT_FOLDER:-"custom-recipes"}

# Check if Containerfile snippets exist
if [[ ! -d "snippets" ]]; then
  echo "No snippets were found!"
  exit 1
fi

# Set some variables used in the Containerfile snippets
OS_VERSION="c9s"
if [[ $BASE_IMAGE == *cuda* ]]; then CUDA='cuda-'; else CUDA=''; fi
if [[ $PYTHON_VERSION == 'py39' ]]; then PYTHON_FOLDER='python3.9'; fi
if [[ $PYTHON_VERSION == 'py311' ]]; then PYTHON_FOLDER='python3.11'; fi

# Prepare temporary folder to create the build source
FOLDER_NAME=${CUDA}${IMAGE_IDE_NAME}-${APP_BUNDLE_NAME}-${OS_VERSION}-${PYTHON_VERSION}_${RELEASE}_${DATE}
mkdir -p $EXPORT_FOLDER/$FOLDER_NAME

# Assemble the Containerfile by concatenating snippets
CONTAINERFILE_CONTENT=""

# Add Builder image before the main one
if [[ -d snippets/bundles/$APP_BUNDLE_FOLDER/builder ]]; then
  if [[ -d snippets/bundles/$APP_BUNDLE_FOLDER/builder/files ]]; then
    cp -r snippets/bundles/$APP_BUNDLE_FOLDER/builder/files/* $EXPORT_FOLDER/$FOLDER_NAME/
  fi
  CONTAINERFILE_CONTENT+=$(envsubst '${BASE_IMAGE}' < snippets/bundles/$APP_BUNDLE_FOLDER/builder/builder.snippet)$'\n\n'
fi

# Start with the base image and labels
CONTAINERFILE_CONTENT+=$(envsubst '${BASE_IMAGE} ${IMAGE_TYPE} ${CUDA} ${IMAGE_IDE_NAME} ${APP_BUNDLE_NAME} ${OS_VERSION} ${PYTHON_VERSION} ${RELEASE} ${DATE}' < snippets/base/base.snippet)$'\n\n'

# Add OS packages
if [[ -d snippets/bundles/$APP_BUNDLE_FOLDER/os ]]; then
  cp snippets/bundles/$APP_BUNDLE_FOLDER/os/os-packages.txt $EXPORT_FOLDER/$FOLDER_NAME/
  CONTAINERFILE_CONTENT+=$(cat "snippets/bundles/os_packages.snippet")$'\n\n'
fi

# Add custom Containerfile code
if [[ -d snippets/bundles/$APP_BUNDLE_FOLDER/custom ]]; then
  if [[ -d snippets/bundles/$APP_BUNDLE_FOLDER/custom/files ]]; then
    cp -r snippets/bundles/$APP_BUNDLE_FOLDER/custom/files/* $EXPORT_FOLDER/$FOLDER_NAME/
  fi
  CONTAINERFILE_CONTENT+=$(cat "snippets/bundles/$APP_BUNDLE_FOLDER/custom/custom.snippet")$'\n\n'
fi

# Add Python packages
if [[ -d snippets/bundles/$APP_BUNDLE_FOLDER/$PYTHON_VERSION ]]; then
  cp snippets/bundles/$APP_BUNDLE_FOLDER/$PYTHON_VERSION/requirements.txt $EXPORT_FOLDER/$FOLDER_NAME/
  CONTAINERFILE_CONTENT+=$(envsubst '${PYTHON_FOLDER}' < snippets/bundles/python_packages.snippet)$'\n\n'
fi

# IDE or Runtime code
if [[ -d snippets/ides/$IMAGE_IDE_FOLDER/files ]]; then
  cp -r snippets/ides/$IMAGE_IDE_FOLDER/files/* $EXPORT_FOLDER/$FOLDER_NAME/
fi
if [[ -d snippets/ides/$IMAGE_IDE_FOLDER/os ]]; then
  mkdir -p $EXPORT_FOLDER/$FOLDER_NAME/os-ide/
  cp -r snippets/ides/$IMAGE_IDE_FOLDER/os/os-packages.txt $EXPORT_FOLDER/$FOLDER_NAME/os-ide/
  CONTAINERFILE_CONTENT+=$(cat "snippets/ides/os_packages_ide.snippet")$'\n\n'
fi
CONTAINERFILE_CONTENT+=$(envsubst '${CUDA} ${PYTHON_FOLDER} ${RELEASE}' < snippets/ides/$IMAGE_IDE_FOLDER/$IMAGE_IDE_NAME.snippet)$'\n\n'

# Jupyter-specific additional code
if [[ $IMAGE_IDE_NAME == 'jupyter' ]]; then 
  cp snippets/bundles/$APP_BUNDLE_FOLDER/$PYTHON_VERSION/requirements-jupyter.txt $EXPORT_FOLDER/$FOLDER_NAME/
fi

echo "$CONTAINERFILE_CONTENT" > $EXPORT_FOLDER/$FOLDER_NAME/Containerfile

tw_echo "Your recipe has been saved to $EXPORT_FOLDER/$FOLDER_NAME." 1
echo
tw_echo "You can build this image with:" 1
tw_echo "cd $EXPORT_FOLDER/$FOLDER_NAME" 1
tw_echo "podman build -t workbench-images:${CUDA}${IMAGE_IDE_NAME}-${APP_BUNDLE_NAME}-${OS_VERSION}-${PYTHON_VERSION}_${RELEASE}_${DATE} ." 1
